File: 백준/Silver/1926. 그림/그림.cc
#include <iostream>
#include <queue>

using namespace std;
using pii=pair<int, int>;

int n, m;
int arr[501][501];
int visited[501][501];
int dx[4] = {1, -1, 0, 0};
int dy[4] = {0, 0, 1, -1};
int ans, sum, cnt;

void bfs(int y, int x) {
	queue<pii> q;
	q.push({y, x});
	visited[y][x] = 1;
	sum = 0;

	while (!q.empty()) {
		y = q.front().first;
		x = q.front().second;
		q.pop();
		sum++;

		for (int i = 0; i < 4; i++) {
			int ny = y + dy[i];
			int nx = x + dx[i];

			if (ny >= 0 && nx >= 0 && ny < n && nx < m) {
				if (!visited[ny][nx] && arr[ny][nx]) {
					visited[ny][nx] = 1;
					q.push({ny, nx});
				}
			}
		}
	}

	ans = max(ans, sum);
}

int main() {
	ios_base::sync_with_stdio(0);cin.tie(0);
	cin >> n >> m;

	for (int i =0 ; i < n; i++) {
		for (int j = 0; j < m; j++) {
			cin >> arr[i][j];
		}
	}

	for (int i = 0; i < n; i++) {
		for (int j = 0; j < m; j++) {
			if (visited[i][j] || !arr[i][j])
				continue;

			bfs(i, j);
			cnt++;
		}
	}

	cout << cnt << "\n";
	cout << ans;

	return 0;
}

File: 백준/Silver/1926. 그림/그림.py
# bfs

import sys
from collections import deque

def bfs(i, j):
    dx = [1, -1, 0, 0]
    dy = [0, 0, 1, -1]
    w = 0
    q = deque()
    q.append((i, j))
    arr[i][j] = 0
    w += 1

    while q:
        x, y = q.popleft()

        for i in range(4):
            nx = x + dx[i]
            ny = y + dy[i]
            if 0 <= nx < n and 0 <= ny < m and arr[nx][ny] == 1:
                q.append((nx, ny))
                arr[nx][ny] = 0
                w += 1
    return w

input = sys.stdin.readline
n, m = map(int, input().split())
arr = []
cnt = 0
ans = 0
for i in range(n):
    ln = list(map(int, input().split()))
    arr.append(ln)

for i in range(n):
    for j in range(m):
        if arr[i][j] == 1:
            cnt += 1
            ans = max(ans, bfs(i, j))

print(cnt)
print(ans)